assets/mdk/proof-dependencies.mdk

```diff
@@ -0,0 +1,87 @@
+
+With v4.4.0[^1] Dafny will include features to detect problems like these, using a notion of _proof dependency_. A proof dependency is a portion of a program that needed to be taken into account in order to construct a proof that the program is correct.
+
+[^1]: Or any nightly build after 2023-10-24.
+
+Let's use the term _property_ more formally from here on to refer a particular statement about a Dafny program that we want to prove. This might arise from an `assert` statement or an `ensures` clause, for example. To legitimately prove that such a property holds of the program, it must be the case that the proof takes the property itself into account. If it's possible to construct a proof that the property holds _without_ considering what it is we're ultimately trying to prove, we say that we've proved it _vacuously_. This can occur because the assumptions in scope at that program location are contradictory, or that that location is unreachable. Contradictory specifications are particularly easy to construct by accident when writing assumed specifications about external or yet-to-be-written code. Consider the following example.
+
+<!-- inline-dafny vacuous/Vacuous -->
+
+Here, the `Find` method has a roughly reasonable-looking (if incomplete) specification. However, it turns out that the `ensures` clause is false if `start` and `end` are the same or adjacent indices, and `CallFind` passes in adjacent indices. Therefore, it's possible to prove a false assertion after the call.
+
+Proof dependencies can also identify unspecified portions of an implementation. If it's possible to prove a program correct without taking into account certain implementation statements, then those statements are unconstrained by the verification. They could be changed arbitrarily without causing verification to fail. In the following example of a binary search implementation, Dafny verifies only the absence of undefined behavior. In particular, it doesn't prove that the return value correctly corresponds to what you'd expect from a binary search implementation. Therefore, any of the return statements could be modified without causing a verification failure.
+
+<!-- inline-dafny unspecified/BinarySearch -->
+
+As a final example, proof dependency analysis can also detect preconditions, assumptions, or intermediate assertions that are unnecessary for constructing a proof of correctness. These redundant specifications may indicate a mistake, such that they would not be redundant if written correctly. Or it may be possible to remove them, simplifying the program and proof process. In the following example, the precondition (which one might think would be required to make the code well-defined) turns out to be unnecessary.
```

Comment:
```suggestion
As a final example, proof dependency analysis can also detect preconditions, assumptions, or intermediate assertions that are unnecessary for constructing a proof of correctness. These redundant specifications may indicate a mistake, such that they would not be redundant if written correctly. Or it may be possible to remove them, simplifying the program and proof process. In the following example, the precondition (which one might think would be required to make the code well-defined) turns out to be unnecessary, and why it's unnecessary is left to you as a thought exercise: Do we really need the array to be sorted to perform a binary search? What is this method returning anyway?.
```